docs/DOCUMENTATION_GUIDELINES.md

```diff
@@ -0,0 +1,227 @@
+
+### TLDR
+
+0. Have the goal of enforcing `#![deny(missing_docs)]` mentally, even if it is not enforced by the compiler 🙈.
+1. Start with a single, clear and concise sentence. Follow up with more context, after a newline, if needed.
+2. Use examples as much as reasonably possible.
+3. Use links as much as possible.
+4. Think about context. If you are in need of explaining a lot of foreign topics in documenting a trait that should not explicitly depend on them, you have likely not designed it properly.
+
+
+### Other Guidelines
+
+The above 5 guidelines must always be respected in docs, as long as reasonable possible.
+
+The following are a set of notes that may not necessarily hold at all circumstances. Consider them as opinionated guidelines that you may or may not abide by.
+
+
+#### Document Through Code
+
+You can certainly do parts of your documentation through your well-organized, properly-named code. That's all true, but:
+
+* This is hard to scale at a project the size of Polkadot/Substrate. Moreover, things like examples, errors and panics cannot be documented via just code.
+
+* Your written documents should *complement* the code, not *repeat* it. Put bluntly, if you end up writing this, you are likely doing it wrong:
+
+ ```rust
+ /// Sends request and handles the response.
+ trait SendRequestAndHandleResponse {
+
+ }
+ ```
+
+ Because the document is adding no extra information and you are better of without it.
+
+
+#### Formatting Matters
+
+The way you format your documents (newlines, heading and so on) do matter! Consider the below examples:
+
+```rust
+/// This function works with input u32 x and multiplies it by two. If
+/// we optimize the other variant of it, we would be able to achieve more
+/// efficiency but I have to think about it. Probably can panic if the input
+/// overflows u32.
+fn multiply_by_2(x: u32) -> u32 { .. }
+```
+
+```rust
+/// Multiplies an input of type [`u32`] by two.
+///
+/// # Panics
+///
+/// Panics if the input overflows.
+///
+/// # Complexity
+///
+/// Is implemented using some algorithm that yields complexity of O(1).
+// More efficiency can be achieved if we improve this via such and such.
+fn multiply_by_2(x: u32) -> u32 { .. }
+```
+
+They are both roughly conveying the same set of facts, but one is substantially kinder on the eye. Especially for traits and type that you can foresee will be seen/used a lot, try and write the better version!
+
+Similarly, make sure your comments are wrapped at 100 characters line-width (as defined by our [`rustfmt.toml`](../rustfmt.toml)), no **more and no less**! The more is fixed by `rustfmt` and our CI, but if you (some some unknown reason) wrap your lines at 59 characters, it will pass the CI, and it will not look good 🫣.
```

Comment:
```suggestion
Similarly, make sure your comments are wrapped at 100 characters line-width (as defined by our [`rustfmt.toml`](../rustfmt.toml)), no **more and no less**! The more is fixed by `rustfmt` and our CI, but if you (for some unknown reason) wrap your lines at 59 characters, it will pass the CI, and it will not look good 🫣.
```